File: corpus/cs2/0180-dna-dinossaurico/solutions/a-optimal-divisors/optimal_divisors.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int main() {
    ios_base::sync_with_stdio(false);
    cin.tie(NULL);
    string s;
    if (!(cin >> s)) return 0;
    int n = s.length();
    
    vector<int> divisors;
    for (int i = 1; i * i <= n; i++) {
        if (n % i == 0) {
            divisors.push_back(i);
            if (i * i != n) {
                divisors.push_back(n / i);
            }
        }
    }
    sort(divisors.begin(), divisors.end());
    
    for (int k : divisors) {
        int d = n / k;
        bool valid = true;
        for (int i = 0; i < d / 2; i++) {
            for (int j = 0; j < k; j++) {
                if (s[i * k + j] != s[(d - 1 - i) * k + j]) {
                    valid = false;
                    break;
                }
            }
            if (!valid) break;
        }
        if (valid) {
            cout << d << "\n";
            return 0;
        }
    }
    return 0;
}

File: corpus/cs2/0180-dna-dinossaurico/solutions/a-optimal-divisors/optimal_divisors.py
import sys

def solve():
    s = sys.stdin.read().split()
    if not s:
        return
    s = s[0]
    n = len(s)
    
    divisors = []
    for i in range(1, int(n**0.5) + 1):
        if n % i == 0:
            divisors.append(i)
            if i * i != n:
                divisors.append(n // i)
    divisors.sort()
    
    for k in divisors:
        d = n // k
        valid = True
        for i in range(d // 2):
            if s[i * k : (i + 1) * k] != s[(d - 1 - i) * k : (d - i) * k]:
                valid = False
                break
        if valid:
            print(d)
            return

if __name__ == '__main__':
    solve()
